src/hotspot/share/opto/opaquenode.cpp

```diff
@@ -102,6 +103,25 @@ const Type* Opaque4Node::Value(PhaseGVN* phase) const {
   return phase->type(in(1));
 }
 
+const Type* OpaqueInitializedAssertionPredicateNode::Value(PhaseGVN* phase) const {
+  return phase->type(in(1));
+}
+
+Node* OpaqueInitializedAssertionPredicateNode::Identity(PhaseGVN* phase) {
```

Comment:
Opaque4 is removed by macro expansion, right? But the new one is removed after loop opts.. So there's a change in behavior. What's the rationale for making that change?

Reply:
That's correct. I've originally had these nodes as macro nodes as well. But concepttionally, we want to get these nodes to be removed and the Initialized Assertion Predicates folded once we know that we no longer split loops (i.e. in post loop IGVN). I think it's easier to register them for this post loop IGVN run since we don't really expand the nodes to anything - they are just removed during expansion.

I'm not entirely sure though what the original reason was to go with a macro expansion removal instead of a post loop IGVN removal for `Opaque4` nodes. Do you remember?

Reply:
> But concepttionally, we want to get these nodes to be removed and the Initialized Assertion Predicates folded once we know that we no longer split loops (i.e. in post loop IGVN).

I don't think that's quite correct. Any round of igvn could cause the bounds of a counted loop to change in a way that conflicts with the types captured in the `CastII`/`ConvI2L` nodes. I think that's true even after loop optimizations are over. As a consequence, we want the Assertion Predicates to fold as late as possible. 

That's poorly tested currently because we emit the predicates in compiled code for debug builds so, in practice, we never really remove them.

As part of this change, I wouldn't change that behavior. That seems risky.

Reply:
I see your point and agree with it. While Template Assertion Predicates should be removed after loop opts are over (no more splitting possible and thus no creation of new Initialized Assertion Predicates) we should indeed delay the removal of the Initialized Assertion Predicates to keep the graph sane. 

Ideally, they should be removed in the very last IGVN round. But there is currently no such dedicated "last IGVN round". Doing the removal in macro expansion is probably fine and as you've stated, it does not change the current behavior.

I've pushed an update doing the change back to macro node.